File: ta_lib/strategies/filter/src/tii.rs
use base::{Filter, OHLCVSeries};
use core::Series;
use trend::tii;

const TII_NEUTRALITY: f32 = 50.0;

pub struct TIIFilter {
    major_period: usize,
    minor_period: usize,
    threshold: f32,
}

impl TIIFilter {
    pub fn new(major_period: f32, minor_period: f32, threshold: f32) -> Self {
        Self {
            major_period: major_period as usize,
            minor_period: minor_period as usize,
            threshold,
        }
    }
}

impl Filter for TIIFilter {
    fn lookback(&self) -> usize {
        std::cmp::max(self.major_period, self.minor_period)
    }

    fn apply(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let tii = tii(&data.close, self.major_period, self.minor_period);

        (
            tii.sgt(TII_NEUTRALITY + self.threshold),
            tii.slt(TII_NEUTRALITY + self.threshold),
        )
    }
}

File: strategy/filter/tii.py
from dataclasses import dataclass

from core.models.parameter import Parameter, RandomParameter
from strategy.filter.base import BaseFilter, FilterType


@dataclass(frozen=True)
class TIIFilter(BaseFilter):
    type: FilterType = FilterType.Tii
    major_period: Parameter = RandomParameter(55.0, 65.0, 5.0)
    minor_period: Parameter = RandomParameter(25.0, 40.0, 5.0)
    threshold: Parameter = RandomParameter(0.0, 5.0, 1.0)
